### PR TITLE
Issue 4/ put-delete-evict transactions

### PR DESCRIPTION
**Pull request by naomijub**

### CODE REVIEW

File: src/gxtdb/adapters/json.clj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+(ns gxtdb.adapters.json
+  (:require [gxtdb.utils :as utils]))
+
+(defn value-record->edn [record]
+  (let [kind (:kind record)
+        key  (-> kind keys first)]
```
Comment (evaporei) [inline src/gxtdb/adapters/json.clj:6]: Is `:kind` always a list with just one item?

File: src/gxtdb/adapters/json.clj [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    (case key
```
Comment (evaporei) [inline src/gxtdb/adapters/json.clj:7]: There's no need for `nil` handling here right?

File: src/gxtdb/adapters/json.clj [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+      :number-value (:number-value kind)
+      :bool-value (:bool-value kind)
+      :string-value (-> kind :string-value str)
+      :list-value (->> kind :list-value :values (mapv value-record->edn))
+      :struct-value (->> kind :struct-value :fields (map (fn [[k v]] {(utils/str->keyword k) (value-record->edn v)})) (reduce into {}))
```
Comment (evaporei) [inline src/gxtdb/adapters/json.clj:12]: I'm sorry I don't remember the context for the `-value` suffixes, but is this inherited from XTDB? I tried searching in the [docs](https://docs.xtdb.com/language-reference/datalog-queries/) but didn't find anything.

Anyhow, if it's on our side, do you think we could do a more elegant solution where the value could be found in one key? And even if we need different handling for each type/kind, we could just look at a keyword describing that.

Eg:
```clojure
{:kind :number
 :value 145}
```

File: test/gxtdb/adapters/tx_log_test.clj [hunk 4]
```diff
@@ -0,0 +1,46 @@
+(ns gxtdb.adapters.tx-log-test
+  #_{:clj-kondo/ignore [:refer-all]}
+  (:require [clojure.test :refer :all]
+            [gxtdb.utils :as utils]
+            [gxtdb.adapters.tx-log :as tx-log]))
+
+(def var->put-tx {:put {:xt-id "id 2", :document {:fields {"painting ids" {:kind {:list-value {:values [{:kind {:number-value 2.0}}, {:kind {:number-value 3.0}} {:kind {:number-value 76.0}} {:kind {:number-value 3.0}}]}}}, "age" {:kind {:number-value 200.0}}, "is-artist?" {:kind {:bool-value true}}, "full name" {:kind {:struct-value {:fields {"last" {:kind {:string-value "picasso"}}, "first" {:kind {:string-value "pablo"}}}}}}, "name" {:kind {:string-value "pablo picasso"}}}}}})
```
Comment (evaporei) [inline test/gxtdb/adapters/tx_log_test.clj:7]: Maybe this is intentional, but do we have a linter/formatter for the project? I think I saw some `clj-kondo`.

Perhaps this map could be better read/understood if that's applied in the tests too :)

File: src/gxtdb/service.clj [hunk 5]
```diff
@@ -43,35 +45,29 @@
 
 (deftype GrpcAPI [xtdb-node]
   api/Service
+  (SubmitTx [_this {{:keys [tx-ops]} :grpc-params}]
+    (let [tx-log (tx-log-adapter/proto->tx-log tx-ops)
+          xt-response (xt/submit-tx xtdb-node tx-log)]
```
Comment (evaporei) [inline src/gxtdb/service.clj:50]: What happens if any of these throw an error? Does the status get set somewhere?

**Review by evaporei — APPROVED** [review]
I've made a few comments, but you have more context than I :shipit:

File: src/gxtdb/adapters/json.clj (hunk 1 above)
Comment (naomijub) [reply src/gxtdb/adapters/json.clj:6]: Kind is alway a map with one item:
`{:kind {:string-value ""}}`, `{:kind {:list-value [...]}}`, `{:kind {:struct-value {...}}}`, etc.

This is generated from the protoc clojure parser.

File: src/gxtdb/adapters/json.clj (hunk 2 above)
Comment (naomijub) [reply src/gxtdb/adapters/json.clj:7]: Added this to be beginner friendly, do you think we should remove?
As the project has two beginners in Clojure

File: src/gxtdb/adapters/json.clj (hunk 3 above)
Comment (naomijub) [reply src/gxtdb/adapters/json.clj:12]: this is generated from protojure, clojure protoc parser. I did not plan on changing the parsing library. Not now at least. We can open an issue for this later

File: test/gxtdb/adapters/tx_log_test.clj (hunk 4 above)
Comment (naomijub) [reply test/gxtdb/adapters/tx_log_test.clj:7]: We do (cljfmt and kibit in CI), I do use kondo and eastwood locally. However, this is generated from protojure

File: src/gxtdb/service.clj (hunk 5 above)
Comment (naomijub) [reply src/gxtdb/service.clj:50]: Will add a test for that, but the proto library should capture the errors

Comment (naomijub) [reply src/gxtdb/service.clj:50]: Could not test this, but manually testings I confim that grpc interceptor caught it and propagated as an error

Comment (naomijub) [reply src/gxtdb/service.clj:50]: https://github.com/naomijub/gXTDB/issues/16

**Pull request merged**